File: examples/school/java/src/main/java/com/example/app/Context.java
package com.example.app;

import com.example.school.TokenIssuer;

/** The Caller is Named before the Story Starts. */
public final class Context {

  private Context() {
  }

  /**
   * RequireProvenCaller Names the Caller before the Story Starts.
   * A Handler behind this Reads request.caller() and Trusts it.
   */
  public static Telling requireProvenCaller(TokenIssuer tokens, Telling tell) {
    return request -> tell.tell(request.namedFor(tokens.readAccessToken(request.token())));
  }
}

File: examples/school/go/app/context.go
package app

import (
	"github.com/cangrejometralleta/OneTwoThree/examples/school/go/school"
	"github.com/cangrejometralleta/OneTwoThree/examples/school/go/transport"
)

// RequireProvenCaller Names the Caller before the Story Starts.
// A Handler behind this Reads req.Caller and Trusts it.
func RequireProvenCaller(tokens school.TokenIssuer, tell Telling) Telling {
	return func(req transport.Request) (any, error) {
		caller, err := tokens.ReadAccessToken(req.Token)
		if err != nil {
			return nil, err
		}

		req.Caller = caller

		return tell(req)
	}
}
